src/providers.rs

```diff
@@ -1,406 +1,259 @@
-pub const BLOCKPI_BASE_MAINNET_HOSTNAME: &str = "base.blockpi.network";
-pub const PUBLICNODE_BASE_MAINNET_HOSTNAME: &str = "base-rpc.publicnode.com";
-pub const ALCHEMY_OPT_MAINNET_HOSTNAME: &str = "opt-mainnet.g.alchemy.com";
-pub const BLOCKPI_OPTIMISM_MAINNET_HOSTNAME: &str = "optimism.blockpi.network";
-pub const PUBLICNODE_OPTIMISM_MAINNET_HOSTNAME: &str = "optimism-rpc.publicnode.com";
-pub const LLAMA_ETH_MAINNET_HOSTNAME: &str = "eth.llamarpc.com";
-pub const LLAMA_ARBITRUM_ONE_HOSTNAME: &str = "arbitrum.llamarpc.com";
-pub const LLAMA_BASE_MAINNET_HOSTNAME: &str = "base.llamarpc.com";
-pub const LLAMA_OPTIMISM_MAINNET_HOSTNAME: &str = "optimism.llamarpc.com";
+const ALCHEMY_HEADERS: &[ConstHeader] = &[ConstHeader {
+    name: "Authorization",
+    value: "Bearer {API_KEY}",
+}];
 
-// Limited API credentials for local testing.
-// Use `dfx canister call evm_rpc updateProvider ...` to pass your own keys.
-pub const ALCHEMY_ETH_MAINNET_CREDENTIAL: &str = "/v2/zBxaSBUMfuH8XnA-uLIWeXfCx1T8ItkM";
-pub const ALCHEMY_ETH_SEPOLIA_CREDENTIAL: &str = "/v2/Mbow19DWsfPXiTpdgvRu4HQq63iYycU-";
-pub const ALCHEMY_ARBITRUM_ONE_CREDENTIAL: &str = "/v2";
-pub const ALCHEMY_BASE_MAINNET_CREDENTIAL: &str = "/v2";
-pub const ALCHEMY_OPTIMISM_MAINNET_CREDENTIAL: &str = "/v2";
-pub const BLOCKPI_ETH_MAINNET_CREDENTIAL: &str = "/v1/rpc/0edc81e20be23ddff051f61a97bb457ec7284a58";
-pub const BLOCKPI_ETH_SEPOLIA_CREDENTIAL: &str = "/v1/rpc/1fe987fddded17db50862311720ff444991d4dab";
-pub const BLOCKPI_ARBITRUM_ONE_CREDENTIAL: &str =
-    "/v1/rpc/a8b89a41d2a341e32ee7aefcb20820a7cbb65f35";
-pub const BLOCKPI_BASE_MAINNET_CREDENTIAL: &str =
-    "/v1/rpc/bd458bf9f28ed45c77823814a937c812d2efd260";
-pub const BLOCKPI_OPTIMISM_MAINNET_CREDENTIAL: &str =
-    "/v1/rpc/d54bfe59299d56b0cbb8b3c69bd122f4ab5ac654";
+pub const PROVIDERS: &[Provider] = &[
+    Provider {
+        provider_id: 0,
+        chain_id: ETH_MAINNET_CHAIN_ID,
+        public_url: Some("https://cloudflare-eth.com/v1/mainnet"),
+        url_pattern: "https://cloudflare-eth.com/v1/mainnet/{API_KEY}",
+        header_patterns: &[],
+        alias: Some(RpcService::EthMainnet(EthMainnetService::Cloudflare)),
+    },
+    Provider {
+        provider_id: 1,
+        chain_id: ETH_MAINNET_CHAIN_ID,
+        public_url: None,
+        url_pattern: "https://rpc.ankr.com/eth/{API_KEY}",
+        header_patterns: &[],
+        alias: Some(RpcService::EthMainnet(EthMainnetService::Ankr)),
+    },
+    Provider {
+        provider_id: 2,
+        chain_id: ETH_MAINNET_CHAIN_ID,
+        public_url: Some("https://ethereum-rpc.publicnode.com"),
+        url_pattern: "https://ethereum-rpc.publicnode.com",
+        header_patterns: &[],
+        alias: Some(RpcService::EthMainnet(EthMainnetService::PublicNode)),
+    },
+    Provider {
+        provider_id: 3,
+        chain_id: ETH_MAINNET_CHAIN_ID,
+        public_url: None,
+        url_pattern: "https://ethereum.blockpi.network/v1/rpc/{API_KEY}",
```

Comment:
Adding this comment to discuss how we could populate these API keys for local testing. We could adjust the [test script](https://github.com/internet-computer-protocol/evm-rpc-canister/blob/main/scripts/e2e) to use environment variables from a gitignored `.env` file, or we could hard-code the local testing API keys. @gregorydemay, do you have a preference for whether we should publicly expose the test keys for convenience?

Reply:
That is a good point. Here are some reasons why the local replica testing has been useful in the past:

* CI coverage for cycles payments
* Monitoring occasional issues with specific combinations of RPC service, method, and chain
* Test coverage for the lightweight [Motoko wrapper library](https://github.com/internet-computer-protocol/evm-rpc-canister/blob/main/lib/motoko/src/lib.mo) included in the repository

We might be able to change this part of the testing workflow but would probably want to include different solutions for the above items. The Motoko wrapper library is probably least important, while the cycles payment CI coverage seems necessary. 

Whether we switch to entirely using the mainnet or continue testing on the local replica, do we want to include the free API keys in the repository? This is currently very convenient for developers learning how to use the EVM RPC canister but isn't a particularly scalable solution. 

If we want to move away from shipping with default API keys, it could make sense to include a script to deploy and configure the keys from a gitignored file. If that sounds reasonable, I'll open a GH issue and address this in a separate PR.

Reply:
> Whether we switch to entirely using the mainnet or continue testing on the local replica, do we want to include the free API keys in the repository? 

I would avoid having any API key in the repository, especially if we need them to test things (meaning that we need them not to be banned or that they still have some quota left)

> This is currently very convenient for developers learning how to use the EVM RPC canister but isn't a particularly scalable solution.

I think this is a double-edge sword. It might work for a few, but as soon as someone max-out the quota (intentionally or not), then that provider that should easily work form the beginning doesn't. I thought that was the whole point of having `public_url` so that people could easily play with the canister without initially needing keys.

> If we want to move away from shipping with default API keys, it could make sense to include a script to deploy and configure the keys from a gitignored file. If that sounds reasonable, I'll open a GH issue and address this in a separate PR.

You have a better overview of the releasing process for the EVM-RPC canister, so feel free to do it. A separate PR sounds good.

Reply:
Tracking in https://github.com/internet-computer-protocol/evm-rpc-canister/issues/264.